tech_soln/real_time_dashboards.rst

```diff
@@ -0,0 +1,422 @@
+ensure there aren't any concurrency bugs where the same rows are written multiple times.
+
+The machinery above which accepts the names of the shards to read and write is necessary
+because only the master has the metadata required to know what the shard pairs are. It has
+its own function to figure that out:
+
+.. code-block:: plpgsql
+
+    -- this function is created on the master
+    CREATE FUNCTION colocated_shard_placements(left_table REGCLASS, right_table REGCLASS)
+    RETURNS TABLE (left_shard TEXT, right_shard TEXT, nodename TEXT, nodeport BIGINT) AS $$
+      SELECT
+        a.logicalrelid::regclass||'_'||a.shardid,
+        b.logicalrelid::regclass||'_'||b.shardid,
+        nodename, nodeport
+      FROM pg_dist_shard a
+      JOIN pg_dist_shard b USING (shardminvalue)
+      JOIN pg_dist_shard_placement p ON (a.shardid = p.shardid)
+      WHERE a.logicalrelid = left_table AND b.logicalrelid = right_table;
+    $$ LANGUAGE 'sql';
+
+Using that metadata, every minute it runs a script which calls ``rollup_1min`` once for
+each pair of shards:
+
+.. code-block:: bash
+
+   #!/usr/bin/env bash
+   
+   QUERY=$(cat <<END
+     SELECT * FROM colocated_shard_placements(
+       'http_request'::regclass, 'http_request_1min'::regclass
+     );
+   END
+   )
+   
+   COMMAND="psql -h \$2 -p \$3 -c \"SELECT rollup_1min('\$0', '\$1')\""
+   
+   psql -tA -F" " -c "$QUERY" | xargs -P32 -n4 sh -c "$COMMAND"
+
+.. NOTE::
+
+  There are many ways to make sure the function is called periodically and no answer that
+  works well for every system. If you're able to run cron on the same machine as the
+  master, and assuming you named the above script ``run_rollups.sh``, you can do something
+  as simple as this:
+
+  .. code-block:: bash
+  
+     * * * * * /some/path/run_rollups.sh
+
+The dashboard query from earlier is now a lot nicer:
+
+.. code-block:: sql
+
+  SELECT
+    request_count, success_count, error_count, average_response_time_msec
+  FROM http_request_1min
+  WHERE site_id = 1 AND minute > date_trunc('minute', now()) - interval '5 minutes';
+
+Expiring Old Data
+-----------------
+
+The rollups make queries faster, but we still need to expire old data to avoid unbounded
+storage costs. Once you decide how long you'd like to keep data for each granularity, you
+could easily write a function to expire old data. In the following example, we decided to
+keep raw data for one day and 1-minute aggregations for one month.
+
+.. code-block:: plpgsql
+
+  -- another function for the master
+  CREATE FUNCTION expire_old_request_data() RETURNS void
+  AS $$
+    SET LOCAL citus.all_modification_commutative TO TRUE;
+    SELECT master_modify_multiple_shards(
+      'DELETE FROM http_request WHERE ingest_time < now() - interval ''1 day'';');
+    SELECT master_modify_multiple_shards(
+      'DELETE FROM http_request_1min WHERE ingest_time < now() - interval ''1 month'';');
+  END;
+  $$ LANGUAGE 'sql';
+
```

Comment:
Calling this function gives me error;

ERROR:  COMMIT is not allowed in a SQL function
CONTEXT:  SQL function "expire_old_request_data" during startup

I think it is because master_modify_multiple_shards uses COMMIT